### DISCUSSION TITLE
Recovering point colours from VoxelGrid

### DISCUSSION THREAD

**Discussion by thesmallblondguy** [daavoo/pyntcloud#315] [Q&A]
**Background:** 
I am working on point cloud change detection. As test data I have two small, registered point clouds -- with some small differences -- which I am trying to force to a matching voxel grid, i.e. I want the co-located points to have equal coordinate values. The goal is to easily hide points with matching x- and y-coordinates from the visualisation (without having to do elaborate queries with ranges around each point). After trying a number of approaches I came across PyntCloud. The  `get_sample("voxelgrid_centers",...` seams promising. However, I have been unable to recover the point colours from the cloud. 

**Question:** 
How can I recover (compute for each new point) the point colours? 

I have tried:

```>>> from pyntcloud import PyntCloud
>>> from pyntcloud.structures import VoxelGrid

>>> pynt_pc = PyntCloud.from_instance('open3d', ref)

>>> pynt_pc
PyntCloud
1803204 points with 3 scalar fields
0 faces in mesh
0 kdtrees
0 voxelgrids
Centroid: -0.5171162016466201, -2.3694974504973105, -1.629357217407153
Other attributes:

>>> voxel_size = 0.005

>>> ref_voxel_id = pynt_pc.add_structure("voxelgrid", 
                        size_x=voxel_size, size_y=voxel_size, size_z=voxel_size)

>>> pynt_pc
PyntCloud
1803204 points with 3 scalar fields
0 faces in mesh
0 kdtrees
1 voxelgrids
Centroid: -0.5171162016466201, -2.3694974504973105, -1.629357217407153
Other attributes:

>>> ref_voxel = pynt_pc.structures[ref_voxel_id]

>>> ref_voxel
<pyntcloud.structures.voxelgrid.VoxelGrid at 0x147dbbfd0>

>>> ref_voxel.colors
Out[13]: 
array([[68, 63, 49],
       [66, 63, 45],
       [70, 65, 47],
       ...,
       [79, 77, 63],
       [70, 70, 54],
       [82, 77, 63]], dtype=uint8)

>>> test_cloud = pynt_pc.get_sample("voxelgrid_centers", voxelgrid_id=ref_voxel_id, 
                             as_PyntCloud=False)

test_cloud
Out[15]: 
               x         y         z
0      -1.880635 -2.964475 -1.628443
1      -1.880635 -2.959475 -1.628443
2      -1.880635 -2.954474 -1.628443
3      -1.880635 -2.949474 -1.628443
4      -1.880635 -2.944473 -1.633451
         ...       ...       ...
211403  0.474354 -1.769337 -1.623435
211404  0.474354 -1.764337 -1.628443
211405  0.474354 -1.764337 -1.623435
211406  0.474354 -1.759336 -1.628443
211407  0.474354 -1.759336 -1.623435

[211408 rows x 3 columns]
```

So, as far as I can tell, the PyntCloud object has colours, but I seem to be unable to recover them. Is there a way to achieve this?

**Comment by daavoo** [comment]
There is not currently a direct way of doing this, although there might be workarounds. 

One question I'm not sure if what values do you expect for the voxel colors to have, a mean of the RGB values of the points inside the voxel?

**Reply by thesmallblondguy** [reply]
Thank you for your quick reply! I understand there are different ways to this. The mean value of the colours of the points inside the voxel sounds ok. I understand this approach is used elsewhere as well. Is there a way, for instance, to get the points (or more precisely the colours) to a DataFrame, from where the mean colour values could be computed efficiently and set as the final point colour?
